backend/storage-service/src/jvmMain/kotlin/file/services/SyncthingClient.kt

```diff
@@ -0,0 +1,304 @@
+
+
+@Serializable
+data class SyncthingFolderDevice(
+    val deviceID: String,
+    val encryptionPassword: String = "",
+    val introducedBy: String = ""
+)
+
+class SyncthingClient(
+    val config: SynchronizationConfiguration,
+    val db: DBContext,
+    val fsPath: String
+) {
+    private val httpClient = HttpClient(CIO) {
+        expectSuccess = false
+    }
+
+    private suspend fun readConfig(device: LocalSyncthingDevice): SyncthingConfig {
+        val requestPath = "http://" + device.hostname + "/rest/config"
+        val resp = httpClient.get<HttpResponse>(requestPath) {
+            headers {
+                append("X-API-Key", device.apiKey)
+            }
+        }
+
+        val config = resp.content.toByteArray().toString(Charsets.UTF_8)
+        return defaultMapper.decodeFromString(config)
+    }
+
+    suspend fun writeConfig() {
+        config.devices.forEach { device ->
+            val result = db.withSession { session ->
+                session.sendPreparedStatement(
+                    {
+                        setParameter("localDevice", device.id)
+                    },
+                    """
+                       select id, path, access_type, d.device_id, d.user_id from storage.synchronized_folders as f
+                       inner join storage.user_devices as d on f.user_id = d.user_id
+                       where f.device_id = :localDevice
+                    """
+                )
+            }.rows
+            val oldConfig = readConfig(device)
+
+            val newConfig = oldConfig.copy(
+                devices = result.map { row ->
+                    SyncthingDevice(
+                        deviceID = row.getField(UserDevicesTable.device),
+                        name = row.getField(UserDevicesTable.device)
+                    )
+                } + listOf(SyncthingDevice(deviceID = device.id, name = device.name)),
+                folders = result.map { row ->
+                    SyncthingFolder(
+                        id = row.getField(SynchronizedFoldersTable.id),
+                        label = row.getField(SynchronizedFoldersTable.path).substringAfterLast("/"),
+                        devices = db.withSession { session ->
+                            session.sendPreparedStatement(
+                                {
+                                    setParameter("user", row.getField(UserDevicesTable.user))
+                                },
+                                """
+                                    select device_id
+                                    from storage.user_devices
+                                    where user_id = :user
+                                """
+                            ).rows.map { SyncthingFolderDevice(it.getField(UserDevicesTable.device)) }
+                        },
+                        path = File(fsPath, row.getField(SynchronizedFoldersTable.path)).absolutePath
+                    )
```

Comment:
A few notes:

- Don't run queries in a loop, it is really bad for performance. Run a single query which fetches everything you need. Ideally this will be combined with your query above.
- Because you are running this across `O(n + 1)` database transactions there are no guarantees that what you found in a previous transaction hasn't already been changed.

You need to rewrite this function such that it only does one or two database queries inside a single transaction.